File: python/src/vigilai/ai_diagnostician.py
"""AI Diagnostician for analyzing incidents and determining root causes"""

import json
import os
import time
import boto3
from typing import Optional, Dict, Any, Literal
from dataclasses import dataclass
from .types import Incident


# Type aliases
DiagnosisCategory = Literal[
    'performance',
    'memory_leak',
    'logic_error',
    'resource_exhaustion',
    'configuration',
    'dependency',
    'unknown'
]


@dataclass
class CodeLocation:
    """Code location information"""
    file: str
    line: Optional[int] = None
    function: Optional[str] = None


@dataclass
class Diagnosis:
    """Diagnosis result from AI analysis"""
    incident_id: str
    root_cause: str
    explanation: str
    category: DiagnosisCategory
    confidence: float  # 0-1
    timestamp: int
    code_location: Optional[CodeLocation] = None
    suggested_fix: Optional[str] = None


class AIDiagnostician:
    """AI Diagnostician class for root cause analysis using Amazon Bedrock"""
    
    DEFAULT_TIMEOUT = 30  # 30 seconds
    DEFAULT_MODEL = 'anthropic.claude-3-sonnet-20240229-v1:0'
    
    def __init__(
        self,
        api_key: Optional[str] = None,  # Kept for backward compatibility but not used
        api_endpoint: Optional[str] = None,  # Kept for backward compatibility but not used
        timeout: Optional[int] = None,
        model: Optional[str] = None
    ):
        self.timeout = timeout or self.DEFAULT_TIMEOUT
        self.model = model or self.DEFAULT_MODEL
        
        # Initialize Bedrock client with credentials from environment
        self.client = boto3.client(
            'bedrock-runtime',
            region_name=os.getenv('AWS_REGION', 'us-east-1'),
            aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
            aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY')
        )
    
    def diagnose(self, incident: Incident) -> Diagnosis:
        """Diagnose an incident using Amazon Bedrock AI"""
        try:
            # Build the diagnosis prompt
            prompt = self._build_diagnosis_prompt(incident)
            
            # Call Bedrock API
            response = self.client.invoke_model(
                modelId=self.model,
                body=json.dumps({
                    'anthropic_version': 'bedrock-2023-05-31',
                    'max_tokens': 2000,
                    'messages': [{
                        'role': 'user',
                        'content': prompt
                    }],
                    'temperature': 0.3  # Lower temperature for more focused responses
                })
            )
            
            # Parse the response
            response_body = json.loads(response['body'].read())
            diagnosis = self._parse_diagnosis_response(incident.id, response_body)
            return diagnosis
            
        except Exception as e:
            # Log error but don't block incident creation
            print(f"Bedrock AI diagnosis failed: {str(e)}")
            return self._create_generic_diagnosis(incident)
    
    def _build_diagnosis_prompt(self, incident: Incident) -> str:
        """Build diagnosis prompt from incident data for Bedrock"""
        parts = [
            'You are an expert software diagnostician. Analyze the incident data and provide a structured diagnosis.',
            '',
            '# Incident Analysis Request',
            '',
            '## Incident Details',
            f'- ID: {incident.id}',
            f'- Type: {incident.type}',
            f'- Severity: {incident.severity}',
            f'- Timestamp: {incident.timestamp}',
            '',
        ]
        
        # Add metrics information
        if incident.metrics:
            parts.append('## Metrics')
            for metric in incident.metrics[:10]:
                tags_str = f' ({json.dumps(metric.tags)})' if metric.tags else ''
                parts.append(f'- {metric.name}: {metric.value}{tags_str}')
            parts.append('')
        
        # Add error information
        if incident.errors:
            parts.append('## Errors')
            for error in incident.errors[:3]:
                parts.append(f'### Error: {error.message}')
                parts.append('```')
                parts.append('\n'.join(error.stack.split('\n')[:10]))
                parts.append('```')
                parts.append('')
        
        # Add context information
        if incident.context:
            parts.append('## Context')
            for key, value in incident.context.items():
                parts.append(f'- {key}: {json.dumps(value)}')
            parts.append('')
        
        parts.extend([
            '## Required Output Format',
            'Provide your diagnosis in the following JSON format:',
            '```json',
            '{',
            '  "rootCause": "Brief description of the root cause",',
            '  "explanation": "Detailed explanation of what is happening",',
            '  "codeLocation": {',
            '    "file": "path/to/file.py",',
            '    "line": 123,',
            '    "function": "function_name"',
            '  },',
            '  "category": "performance|memory_leak|logic_error|resource_exhaustion|configuration|dependency",',
            '  "confidence": 0.85,',
            '  "suggestedFix": "High-level description of how to fix the issue"',
            '}',
            '```',
        ])
        
        return '\n'.join(parts)
    
    def _parse_diagnosis_response(self, incident_id: str, response: Dict[str, Any]) -> Diagnosis:
        """Parse Bedrock AI response into Diagnosis object"""
        try:
            # Extract content from Bedrock response
            content = response.get('content', [{}])[0].get('text', '')
            
            # Try to extract JSON from the response
            import re
            json_match = re.search(r'```json\s*([\s\S]*?)\s*```', content) or \
                        re.search(r'\{[\s\S]*\}', content)
            
            if json_match:
                json_str = json_match.group(1) if json_match.lastindex else json_match.group(0)
                parsed = json.loads(json_str)
                
                # Extract code location if present
                code_location = None
                if parsed.get('codeLocation'):
                    loc = parsed['codeLocation']
                    code_location = CodeLocation(
                        file=loc.get('file', ''),
                        line=loc.get('line'),
                        function=loc.get('function')
                    )
                
                return Diagnosis(
                    incident_id=incident_id,
                    root_cause=parsed.get('rootCause', 'Unable to determine root cause'),
                    explanation=parsed.get('explanation', 'No detailed explanation available'),
                    code_location=code_location,
                    category=self._validate_category(parsed.get('category', 'unknown')),
                    confidence=max(0, min(1, parsed.get('confidence', 0.5))),
                    suggested_fix=parsed.get('suggestedFix'),
                    timestamp=int(time.time() * 1000)
                )
            
            # If no JSON found, try to extract information from text
            return Diagnosis(
                incident_id=incident_id,
                root_cause=content[:200] if content else 'Unable to determine root cause',
                explanation=content or 'AI response could not be parsed',
                category='unknown',
                confidence=0.3,
                timestamp=int(time.time() * 1000)
            )
            
        except Exception as e:
            print(f"Failed to parse Bedrock AI response: {str(e)}")
            return Diagnosis(
                incident_id=incident_id,
                root_cause='Failed to parse AI diagnosis',
                explanation='The AI response could not be parsed into a structured diagnosis',
                category='unknown',
                confidence=0,
                timestamp=int(time.time() * 1000)
            )
    
    def _validate_category(self, category: str) -> DiagnosisCategory:
        """Validate and normalize category"""
        valid_categories = [
            'performance',
            'memory_leak',
            'logic_error',
            'resource_exhaustion',
            'configuration',
            'dependency',
        ]
        
        if category in valid_categories:
            return category  # type: ignore
        
        return 'unknown'
    
    def _create_generic_diagnosis(self, incident: Incident) -> Diagnosis:
        """Create a generic diagnosis when AI fails"""
        root_cause = 'Unable to determine root cause'
        category: DiagnosisCategory = 'unknown'
        explanation = 'AI diagnosis service unavailable. Manual investigation required.'
        
        # Provide basic categorization based on incident type
        if incident.type == 'performance':
            category = 'performance'
            root_cause = 'Performance degradation detected'
            explanation = 'The application is experiencing performance issues. Check response times and resource usage.'
        elif incident.type == 'error':
            category = 'logic_error'
            root_cause = 'Error rate exceeded threshold'
            explanation = 'Multiple errors detected. Review error logs and stack traces for details.'
        elif incident.type == 'resource':
            category = 'resource_exhaustion'
            root_cause = 'Resource usage exceeded threshold'
            explanation = 'System resources (CPU/memory) are being exhausted. Check for memory leaks or inefficient code.'
        
        return Diagnosis(
            incident_id=incident.id,
            root_cause=root_cause,
            explanation=explanation,
            category=category,
            confidence=0,
            timestamp=int(time.time() * 1000)
        )

# Add missing import
import time

File: typescript/src/ai-diagnostician.ts
/**
 * AI Diagnostician for analyzing incidents and determining root causes using Amazon Bedrock
 */

import { BedrockRuntimeClient, InvokeModelCommand } from '@aws-sdk/client-bedrock-runtime';
import { Incident } from './types';

/**
 * Diagnosis category types
 */
export type DiagnosisCategory = 
  | 'performance' 
  | 'memory_leak' 
  | 'logic_error' 
  | 'resource_exhaustion'
  | 'configuration'
  | 'dependency'
  | 'unknown';

/**
 * Code location information
 */
export interface CodeLocation {
  file: string;
  line?: number;
  function?: string;
}

/**
 * Diagnosis result from AI analysis
 */
export interface Diagnosis {
  incidentId: string;
  rootCause: string;
  explanation: string;
  codeLocation?: CodeLocation;
  category: DiagnosisCategory;
  confidence: number; // 0-1
  suggestedFix?: string;
  timestamp: number;
}

/**
 * AI Diagnostician configuration
 */
export interface AIDiagnosticianConfig {
  apiKey?: string; // Kept for backward compatibility but not used
  apiEndpoint?: string; // Kept for backward compatibility but not used
  timeout?: number; // milliseconds, default: 30000
  model?: string; // AI model to use
}

/**
 * AI Diagnostician class for root cause analysis using Amazon Bedrock
 */
export class AIDiagnostician {
  private client: BedrockRuntimeClient;
  private config: AIDiagnosticianConfig;
  private readonly DEFAULT_TIMEOUT = 30000; // 30 seconds
  private readonly DEFAULT_MODEL = 'anthropic.claude-3-sonnet-20240229-v1:0';

  constructor(config: AIDiagnosticianConfig = {}) {
    this.config = {
      ...config,
      timeout: config.timeout || this.DEFAULT_TIMEOUT,
      model: config.model || this.DEFAULT_MODEL,
    };

    // Initialize Bedrock client with credentials from environment
    this.client = new BedrockRuntimeClient({
      region: process.env.AWS_REGION || 'us-east-1',
      credentials: {
        accessKeyId: process.env.AWS_ACCESS_KEY_ID || '',
        secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || '',
      },
    });
  }

  /**
   * Diagnose an incident using Amazon Bedrock AI
   */
  async diagnose(incident: Incident): Promise<Diagnosis> {
    try {
      // Build the diagnosis prompt
      const prompt = this.buildDiagnosisPrompt(incident);

      // Call Bedrock API
      const command = new InvokeModelCommand({
        modelId: this.config.model,
        body: JSON.stringify({
          anthropic_version: 'bedrock-2023-05-31',
          max_tokens: 2000,
          messages: [{
            role: 'user',
            content: prompt,
          }],
          temperature: 0.3, // Lower temperature for more focused responses
        }),
      });

      const response = await this.client.send(command);
      
      // Parse the response
      const responseBody = JSON.parse(new TextDecoder().decode(response.body));
      const diagnosis = this.parseDiagnosisResponse(incident.id, responseBody);
      return diagnosis;

    } catch (error) {
      // Log error but don't block incident creation
      console.error('Bedrock AI diagnosis failed:', error instanceof Error ? error.message : 'Unknown error');
      return this.createGenericDiagnosis(incident);
    }
  }

  /**
   * Build diagnosis prompt from incident data for Bedrock
   */
  private buildDiagnosisPrompt(incident: Incident): string {
    const parts: string[] = [
      'You are an expert software diagnostician. Analyze the incident data and provide a structured diagnosis.',
      '',
      '# Incident Analysis Request',
      '',
      `## Incident Details`,
      `- ID: ${incident.id}`,
      `- Type: ${incident.type}`,
      `- Severity: ${incident.severity}`,
      `- Timestamp: ${new Date(incident.timestamp).toISOString()}`,
      '',
    ];

    // Add metrics information
    if (incident.metrics.length > 0) {
      parts.push('## Metrics');
      incident.metrics.slice(0, 10).forEach(metric => {
        parts.push(`- ${metric.name}: ${metric.value}${metric.tags ? ` (${JSON.stringify(metric.tags)})` : ''}`);
      });
      parts.push('');
    }

    // Add error information
    if (incident.errors.length > 0) {
      parts.push('## Errors');
      incident.errors.slice(0, 3).forEach(error => {
        parts.push(`### Error: ${error.message}`);
        parts.push('```');
        parts.push(error.stack.split('\n').slice(0, 10).join('\n'));
        parts.push('```');
        parts.push('');
      });
    }

    // Add context information
    if (incident.context && Object.keys(incident.context).length > 0) {
      parts.push('## Context');
      Object.entries(incident.context).forEach(([key, value]) => {
        parts.push(`- ${key}: ${JSON.stringify(value)}`);
      });
      parts.push('');
    }

    parts.push('## Required Output Format');
    parts.push('Provide your diagnosis in the following JSON format:');
    parts.push('```json');
    parts.push('{');
    parts.push('  "rootCause": "Brief description of the root cause",');
    parts.push('  "explanation": "Detailed explanation of what is happening",');
    parts.push('  "codeLocation": {');
    parts.push('    "file": "path/to/file.ts",');
    parts.push('    "line": 123,');
    parts.push('    "function": "functionName"');
    parts.push('  },');
    parts.push('  "category": "performance|memory_leak|logic_error|resource_exhaustion|configuration|dependency",');
    parts.push('  "confidence": 0.85,');
    parts.push('  "suggestedFix": "High-level description of how to fix the issue"');
    parts.push('}');
    parts.push('```');

    return parts.join('\n');
  }

  /**
   * Parse Bedrock AI response into Diagnosis object
   */
  private parseDiagnosisResponse(incidentId: string, response: any): Diagnosis {
    try {
      // Extract content from Bedrock response
      const content = response.content?.[0]?.text || '';
      
      // Try to extract JSON from the response
      const jsonMatch = content.match(/```json\s*([\s\S]*?)\s*```/) || 
                       content.match(/\{[\s\S]*\}/);
      
      if (jsonMatch) {
        const jsonStr = jsonMatch[1] || jsonMatch[0];
        const parsed = JSON.parse(jsonStr);

        return {
          incidentId,
          rootCause: parsed.rootCause || 'Unable to determine root cause',
          explanation: parsed.explanation || 'No detailed explanation available',
          codeLocation: parsed.codeLocation,
          category: this.validateCategory(parsed.category),
          confidence: Math.min(Math.max(parsed.confidence || 0.5, 0), 1),
          suggestedFix: parsed.suggestedFix,
          timestamp: Date.now(),
        };
      }

      // If no JSON found, try to extract information from text
      return {
        incidentId,
        rootCause: content.substring(0, 200) || 'Unable to determine root cause',
        explanation: content || 'AI response could not be parsed',
        category: 'unknown',
        confidence: 0.3,
        timestamp: Date.now(),
      };

    } catch (error) {
      console.error('Failed to parse Bedrock AI response:', error);
      return {
        incidentId,
        rootCause: 'Failed to parse AI diagnosis',
        explanation: 'The AI response could not be parsed into a structured diagnosis',
        category: 'unknown',
        confidence: 0,
        timestamp: Date.now(),
      };
    }
  }

  /**
   * Validate and normalize category
   */
  private validateCategory(category: string): DiagnosisCategory {
    const validCategories: DiagnosisCategory[] = [
      'performance',
      'memory_leak',
      'logic_error',
      'resource_exhaustion',
      'configuration',
      'dependency',
    ];

    if (validCategories.includes(category as DiagnosisCategory)) {
      return category as DiagnosisCategory;
    }

    return 'unknown';
  }

  /**
   * Create a generic diagnosis when AI fails
   */
  private createGenericDiagnosis(incident: Incident): Diagnosis {
    let rootCause = 'Unable to determine root cause';
    let category: DiagnosisCategory = 'unknown';
    let explanation = 'AI diagnosis service unavailable. Manual investigation required.';

    // Provide basic categorization based on incident type
    if (incident.type === 'performance') {
      category = 'performance';
      rootCause = 'Performance degradation detected';
      explanation = 'The application is experiencing performance issues. Check response times and resource usage.';
    } else if (incident.type === 'error') {
      category = 'logic_error';
      rootCause = 'Error rate exceeded threshold';
      explanation = 'Multiple errors detected. Review error logs and stack traces for details.';
    } else if (incident.type === 'resource') {
      category = 'resource_exhaustion';
      rootCause = 'Resource usage exceeded threshold';
      explanation = 'System resources (CPU/memory) are being exhausted. Check for memory leaks or inefficient code.';
    }

    return {
      incidentId: incident.id,
      rootCause,
      explanation,
      category,
      confidence: 0,
      timestamp: Date.now(),
    };
  }
}
